### PR TITLE
[ansible] Add Ubuntu 20.04 linux-riscv64 image

### PR DESCRIPTION
**Pull request by luhenry**
It will be used for cross-compilation and testing.

##### Checklist

- [x] commit message has one of the [standard prefixes](https://github.com/adoptium/infrastructure/blob/master/CONTRIBUTING.md#commit-messages)
- [x] [faq.md](https://github.com/adoptium/infrastructure/blob/master/FAQ.md) updated if appropriate
- [ ] other documentation is changed or added (if applicable)
- [ ] playbook changes run through [VPC](https://ci.adoptium.net/view/Tooling/job/VagrantPlaybookCheck/) or [QPC](https://ci.adoptium.net/view/Tooling/job/QEMUPlaybookCheck/) (if you have access)
- [ ] VPC/QPC not applicable for this PR
- [ ] for inventory.yml changes, bastillion/nagios/jenkins updated accordingly

### CODE REVIEW

**Comment by luhenry** [comment]
I'm not sure how to update https://ci.adoptium.net/view/Tooling/job/QEMUPlaybookCheck to include a ubuntu2004 check for riscv

**Review by karianna — APPROVED** [review]

**Comment by luhenry** [comment]
Could you please add the `QEMU-playbook-check` label so that it triggers the check on CI?

**Comment by sxa** [comment]
Thanks - I'll try a run of the new dockerfile on my system :-)

**Review by sxa — COMMENTED** [review]
The docs on how QEMUPlaybook is done is under https://github.com/adoptium/infrastructure/blob/master/docs/Setup-QEMU-Images.md and it's hosted on a specific machine that you'd need access to to be able to do any of the updates as they're not currently automated.

https://github.com/adoptium/infrastructure/blob/master/docs/Setup-RISCV-VMs.md has some info on how the existing RISC-V QPC virtual machine was initially set up.

As you can probably tell the QPC stuff hasn't been updated in a while as the latest is Debian 11/Ubuntu18 so it needs a bit of attention - there's an issue relating to that at https://github.com/adoptium/infrastructure/issues/2121 but to be clear I don't consider the lack of QPC for this version to be a blocker for this PR :-)

Can you clarify the reason for having mercurial-common, python2, and upc explicitly called out? We don't use mercurial any more and I'm a little surprised we'd need to special case the others too so would be good to clarify what they're needed for if we're going to include them.

Also have you been able to test that the github actions stuff being added here work ok?

**Comment by sxa** [comment]
I can't remember if the actions will trigger on the contents of the PR of if the PR has to be merged first, so was wondering if it might be simpler to test it in your own repository if that's feasisble.

Yeah mercurial was useful before openjdk moved itself to GitHub so shouldn't be necessary now as far as I'm aware (especially not on RISC-V) so I'm ok with having that stuff removed.

**Comment by sxa** [comment]
> Thanks - I'll try a run of the new dockerfile on my system :-)

FYI that seemed to work well so nice to see this seems to work - I'll try and kick off a build in the new container image overnight on my machine :-)

**Comment by luhenry** [comment]
I've verified locally that the image allows us to build jdk17u with:
```
podman run --rm -it --platform linux/riscv64 -v /workspace:/workspace -w $(pwd) localhost/adoptopenjdk/ubuntu2004_build_image:linux-riscv64 bash -ex -c "cd build-farm; ./make-adopt-build-farm.sh jdk17u"
```

**Comment by luhenry** [comment]
I've verified the latest changes work for building `jdk20u`:
```
podman run --rm -it --platform linux/riscv64 -v /workspace:/workspace -w $(pwd) localhost/adoptopenjdk/ubuntu2004_build_image:linux-riscv64 bash -ex -c "cd build-farm; bash -x ./make-adopt-build-farm.sh jdk20u"
```

File: ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/adoptopenjdk_install/tasks/main.yml [hunk 1]
```diff
@@ -172,6 +173,36 @@
...
+- name: Install JDK {{ jdk_version }} on Linux-riscv64
+  when:
+    - ansible_architecture == "riscv64" and jdk_version == 19
+    - adoptopenjdk_installed.rc != 0
+  tags: adoptopenjdk_install
+    # Api does not return release information for JDK10
+  block:
+    - name: Download jdk{{ jdk_version }} release (Linux-riscv64)
+      get_url:
+        url: https://ci.adoptium.net/job/build-jdk19-rv64-cross/392/artifact/archive/jdk19u-riscv64-20231104.tar.gz
```
Comment (sxa) [inline ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/adoptopenjdk_install/tasks/main.yml:187]:
```suggestion
        url: https://ci.adoptium.net/userContent/riscv/jdk19u-riscv64-20231107.glib227.tar.gz
```

**Comment by sxa** [comment]
VPC run at https://ci.adoptium.net/view/Tooling/job/VagrantPlaybookCheck/1742 as a final check.

**Comment by luhenry** [comment]
It succeeded to build jdk20u locally with the latest change.

**Comment by sxa** [comment]
Noting that this will currently misbehave on many real boards due to https://github.com/adoptium/adoptium-support/issues/697 unless `--security-opt seccomp=unconfined` is added to the `DOCKER_ARGS` when starting the container.

**Comment by sxa** [comment]
After setting those options to stop the JVM crashing I'm getting a segfault in the compiler during the openjdk build process at the moment which isn't great :'(
```
13:58:44  g++: internal compiler error: Segmentation fault signal terminated program cc1plus
```

**Comment by luhenry** [comment]
> Noting that this will currently misbehave on many real boards due to https://github.com/adoptium/adoptium-support/issues/697 unless --security-opt seccomp=unconfined is added to the DOCKER_ARGS when starting the container.

We can add the flag in https://github.com/adoptium/ci-jenkins-pipelines/pull/836.

**Comment by luhenry** [comment]
> After setting those options to stop the JVM crashing I'm getting a segfault in the compiler during the openjdk build process at the moment which isn't great :'(
> 
> 13:58:44  g++: internal compiler error: Segmentation fault signal terminated program cc1plus

From talking offline, that's when _building_ on a board. I want to push further on building inside docker containers, to free up board for testing. We'll also want to use QEMU for testing for the newer extensions that we are adding support for in OpenJDK (Vector, Vector Crypto, and others). We won't be able to do this testing on boards for the next few years, even though these extensions will be common-place.

**Comment by luhenry** [comment]
I can't reproduce the g++ SIGSEGV when bumping to gcc-10/g++-10 and running on VF2.

**Review by sxa — APPROVED** [review]
I think I'm happy for this to go in now. Noting that jdk19u will likely not be required long term and I'll look at the4 feasibility of publishing a jdk20u build against the Ubuntu 20.04 glibc which can be consumed via the API at which point we can likely clean up the JDK download stuff a bit more for the future.

**Pull request merged**